src/coreclr/jit/assertionprop.cpp

```diff
@@ -3568,6 +3568,20 @@ GenTree* Compiler::optCopyAssertionProp(AssertionDsc*        curAssertion,
         return nullptr;
     }
 
+    // Heuristic: for LclFld prop, don't force the copy or its promoted fields to be in memory.
+    //
+    if (tree->OperIs(GT_LCL_FLD))
+    {
+        if (copyVarDsc->IsEnregisterableLcl() || copyVarDsc->lvPromotedStruct())
```

Comment:
Nit: one almost never wants to call `lvPromotedStruct()` instead of just checking `lvPromoted`. If we start having non-struct promoted variables in the future, we'll want this behavior for them as well.

Orthogonally, this is a bit conservative. If we already know `copyVarDsc` is not going to be enregistered, it should be ok to propagate it, so `lvaGetPromotionType(copyVarDsc) == PROMOTION_TYPE_INDEPENDENT` should in theory be a better check.

Reply:
I saw quite a few more regressions with promoted structs. Will have to look more closely.

Reply:
> I saw quite a few more regressions with promoted structs

I suppose that addresses the `PROMOTION_TYPE_INDEPENDENT` idea. Guessing these could have been from SSA-based optimization not working as well (since promoted structs are excluded from all of them).